### PR TITLE
Update to Starknet 0.13.0

### PR DESCRIPTION
**Pull request by FabijanC**
## Usage related changes

- Update to Starknet 0.13.0

## Development related changes

- Related to updating to JSON-RPC v0.6.0
- Update blockifier to v0.4.0-rc8 (and related dependencies as well)
- One TODO left in the code - the new tx fee errors are complicated so I'm unsure about the extensiveness of the matching
  - if no objections, I'll just remove the TODO and the commented line(s)
- Rename our custom `FeeError` to `MaxFeeZeroError`
  - I thought it would be necessary to introduce a new Fee related error so I thought this renaming was necessary, but it turned out all the tests are passing without the new error; still, I think it's better with this change
- Some tests were failing because now invalid nonce is reported before insufficient balance, so I had to modify the declaration body; these are the tests, both in `get_transaction_receipt_by_hash_integration_tests`:
  - `declare_v1_accepted_with_numeric_entrypoint_offset`
  - `declare_v1_transaction_fails_with_insufficient_max_fee`

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Checked the TODO section in README.md if this PR resolves it
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-server/test_data/rpc/declare_v1.json [hunk 1]
```diff
@@ -1 +1 @@
...
...+DHoCUoaO9s2peMCJamSXrco2yybMtJt9NAUMtOFBX7745e3V238EG2m27Woh2EAYE4oYJVBwSBkDlBEDUcC4oJRRKBDiiBADGtTYeU0pQVT+apK+EhgggwmCEQMQMr0XFBrGzlt7Xo7thbV40Dlshw4n1svg9yDGJdLGFsQc3vvYCPMI9x7mSLaNRe6b69fPzVfXNzd6y/PtG+0eCtyCXsMVgDODY4AJFgIwKAxOBSKYKSex+1petOP67fXV84J1RAC4AIbAmAFDryXrLd+Aowa+IOr0b29f3BasXiLKEVe9UYA4I4ACzathGEgwvDuzcuq+Ck5F3c5ewpuHZa47qo2S1az2Dt259TCZWW6OqaPtD2Wu+sBtWnuBlDYWT3a3+GA//tvobsduBoFYL9sHAdS4hyZjQFwgbQSbOaw3tu5apt1QCFn3itqRRUFM2uRizB1JaCPiDtv92/r20P0pWBW3c0KyGnNtgKJgbmhzUX9FMTmOvp7P3PlUxp31xfRkdXdfkjDh2glT2e5fxc50aQVRWzbWKzny53zQd56M3coH/6az6NqfJmf/tmm6xnG/apcIl9wIBoouJ6rZR8GNkF1PzWZJaMM/JVpt+frIxlwW3KLUErfF9yrF32Wy7W6jur3Wve2obvtpNTVSQLyxAi374ZonF3rJYBm86X2Ntoizk/z3fz85e+e/mmqeeXV5yrreL2b3eusL0CYVtk/i7QcvbFy3r1/RiM9jl1OVdwGTXbyL6WHNWl1dbL0unLfgjUruEa97hKLG7QtNYJ17IzlqjOdO7ipvzmdn/qvxJeK1++rQbQUdtOW3sq3VdFy80HHRnR0XjDkuiAs9F23iuWCh5/r27f8B/fXlzOt2AQA="}, "nonce": "0x0", "signature": ["0xacde09b7b4218575c154dbf9a0598c4e76f13dba493d048e0e79221960d353", "0x1872cdcabaa4c2746741f120a82ca09ced10cd46f62bc2e7c07599aff180f80"], "max_fee": "0x2af56b9ed300", "sender_address": "0xd47d2f1a72351f234dfc42a3f6b5cd6a16b32491daeca9bbb02e9e990e0e12", "version": "0x1", "type": "DECLARE"}
```
Comment (marioiordanov) [inline crates/starknet-server/test_data/rpc/declare_v1.json:1]: why the nonce is changed

Comment (FabijanC) [reply crates/starknet-server/test_data/rpc/declare_v1.json:1]: I think it was wrong all this time, but I guess the sofar the balance was internally checked before the nonce. I mentioned this in the PR description (last bullet).

Comment (FabijanC) [reply crates/starknet-server/test_data/rpc/declare_v1.json:1]: if it's a first tx by a predeployed account (or just an attempted tx by a non-deployed account), I think nonce should be zero

File: crates/starknet/src/starknet/mod.rs [hunk 2]
```diff
@@ -274,13 +275,24 @@ impl Starknet {
...
+                        blockifier::transaction::errors::TransactionFeeError::FeeTransferError { .. }
+                        | blockifier::transaction::errors::TransactionFeeError::MaxFeeTooLow { .. } => Err(
+                            TransactionValidationError::InsufficientMaxFee.into()
+                        ),
+                        blockifier::transaction::errors::TransactionFeeError::MaxFeeExceedsBalance { .. } => Err(
+                            TransactionValidationError::InsufficientAccountBalance.into()
+                        ),
+                        _ => Err(tx_fee_err.into())
+                    },
+                    // blockifier::transaction::errors::TransactionExecutionError::TransactionFeeError { .. } =>
```
Comment (marioiordanov) [inline crates/starknet/src/starknet/mod.rs:294]: what about TransactionFeeError now its going to fall in the default case

Comment (FabijanC) [reply crates/starknet/src/starknet/mod.rs]: As discussed on Discord, I addressed this in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/256/commits/87de6234ab06c098cc7859b5a97f61502907ef00

**Review by marioiordanov — APPROVED** [review]

**Pull request merged**